src/aspire/source/relion.py

```diff
@@ -119,26 +192,63 @@ def __init__(self, filepath, pixel_size=1, B=0, n_workers=-1, ignore_missing_fil
                     voltage=row[0],
                     defocus_u=row[1],
                     defocus_v=row[2],
-                    defocus_ang=row[3],
+                    defocus_ang=row[3] * np.pi / 180,  # degrees to radians
                     Cs=row[4],
                     alpha=row[5],
-                    B=self.B
+                    B=B
                 )
             )
-        filters = SourceFilter(filters, indices=filter_indices)
 
-        offsets = self.df[['rlnOriginX', 'rlnOriginY']].values
-        amplitudes = np.ones(self.n)
-        states = self.df['rlnClassNumber'].values
+        metadata['__filter'] = [filters[i] for i in filter_indices]
 
         ImageSource.__init__(
             self,
-            L=self.L,
-            n=self.n,
-            states=states,
-            filters=filters,
-            offsets=offsets,
-            amplitudes=amplitudes,
-            rots=rots,
-            dtype=self.dtype
+            L=L,
+            n=n,
+            dtype=dtype,
+            metadata=metadata
         )
+
+    def __str__(self):
+        return f'Starfile ({self.n} images of size {self.L}x{self.L})'
+
+    def _images(self, start=0, num=None):
+
+        def load_single_mrcs(filepath, df):
+            arr = mrcfile.open(filepath).data
+            data = arr[df['__mrc_index'] - 1, :, :].T
+
+            if self.L < self._L:
+                data = Image(data).downsample(self.L).asnumpy()
+
+            return df.index, data
+
+        n_workers = self.n_workers
+        if n_workers < 0:
+            n_workers = cpu_count() - 1
+
+        if num is None:
+            num = self.n - start
+        else:
+            num = min(self.n - start, num)
+
+        df = self._metadata[start:num]
+        im = np.empty((self.L, self.L, num))
+
+        groups = df.groupby('__mrc_filepath')
+        n_workers = min(n_workers, len(groups))
+
+        pbar = tqdm(total=self.n)
+        with futures.ThreadPoolExecutor(n_workers) as executor:
```

Comment:
Is it necessary to multithread this? We typically won't be loading all that many images at a time (not more than a few thousand) to process them, so it's likely that we'd want to multithread that outer loop in many cases.

Reply:
This multi-threading brought down the reading of a huge STAR file (and 330GB of associated mrcs files) from several hours down to 20 minutes. This might prove key once we start experimenting with real datasets. Better to leave it in there rather than having to add it later.